helix-core/src/indent.rs

```diff
@@ -186,106 +188,338 @@ pub fn indent_level_for_line(line: RopeSlice, tab_width: usize) -> usize {
     len / tab_width
 }
 
-/// Find the highest syntax node at position.
-/// This is to identify the column where this node (e.g., an HTML closing tag) ends.
-fn get_highest_syntax_node_at_bytepos(syntax: &Syntax, pos: usize) -> Option<Node> {
-    let tree = syntax.tree();
+/// The indent that is added for a single tree-sitter node/a single line.
+/// This is different from the total indent ([IndentResult]) because multiple indents/outdents on the same line don't stack.
+struct AddedIndent {
+    indent: bool,
+    outdent: bool,
+}
+impl AddedIndent {
+    fn new() -> Self {
+        AddedIndent {
+            indent: false,
+            outdent: false,
+        }
+    }
+    /// Combine this [AddedIndent] with other.
+    /// This is intended for indents that apply to the same line.
+    fn combine_with(&mut self, other: &AddedIndent) {
+        self.indent |= other.indent;
+        self.outdent |= other.outdent;
+    }
+}
 
-    // named_descendant
-    let mut node = match tree.root_node().descendant_for_byte_range(pos, pos) {
-        Some(node) => node,
-        None => return None,
-    };
+/// The total indent for some line of code.
+/// This is usually constructed by successively adding instances of [AddedIndent]
+struct IndentResult {
+    /// The total indent (the number of indent levels).
+    /// The string that this results in depends on the indent style (spaces or tabs, etc.)
+    indent: i32,
+}
+impl IndentResult {
+    fn new() -> Self {
+        IndentResult { indent: 0 }
+    }
+    /// Add the given [AddedIndent] to the [IndentResult].
+    /// The [AddedIndent] should be the combination of all the added indents for one line.
+    fn add(&mut self, added: &AddedIndent) {
+        if added.indent && !added.outdent {
+            self.indent += 1;
+        } else if added.outdent && !added.indent {
+            self.indent = self.indent.saturating_sub(1);
+        }
+    }
+    fn as_string(&self, indent_style: &IndentStyle) -> String {
+        indent_style
+            .as_str()
+            .repeat(std::cmp::max(self.indent, 0) as usize)
```

Comment:
```suggestion
            .repeat(0.max(self.indent as usize))
```

Reply:
Definitely looks nicer. I think we still need to take the maximum before casting, otherwise a negative value for `self.indent` (which could happen depending on the indent query) will lead to very weird results